### ISSUE TITLE
Experimental TypeGuard produces invalid results?

### ISSUE THREAD

**Issue by rchiodo** [microsoft/pyright#5814]
After enabling the experimental type guard support, this code errors:

```python
_T = TypeVar("_T")


def is_two_element_tuple(a: tuple[_T, ...]) -> TypeGuard[tuple[_T, _T]]:
    return len(a) == 2


def func1(a: tuple[int, ...]):
    if is_two_element_tuple(a):
        reveal_type(a, expected_text="tuple[int, int]")
    else:
        reveal_type(a, expected_text="tuple[int, ...]")
```

Pyright is saying the type of a is always `tuple[int, ...]`. I found this as part of the experimental run of mypy primer to see if the new typeguard would break anything.

**Comment by rchiodo** [comment]
Experimental run is here:
https://github.com/microsoft/pyright/pull/5813

Note: Just to be clear, the error above doesn't require the changes in that PR, it just requires the 'enableExperimentalFeatures' to be true.

**Comment by erictraut** [comment]
The current implementation determines whether to apply "strict type guard" semantics based on whether the return type is "consistent" with the input type. Here I use the word "consistent" in the sense that it is used in PEP 483. This works well for types that honor subtyping rules, but it falls apart for types that that extend the subtyping rules for purposes of gradual typing. For example, `int` is a subtype of `float`, and `float` is a supertype of `int`. But `int` is not technically a subtype of `Any` because `Any` doesn't follow the rules of subtyping. Instead, we say that `int` is consistent with `Any` and `Any` is consistent with `int`.

`Any` isn't the only type in the Python type system that doesn't follow subtyping rules. This list includes the `...` forms of tuples and the `...` form of `Callable` and `ParamSpec`.

With all of this in mind, I think we need to modify the rules for "strict type guard" semantics. I can think of two different options here.
1. Rather than simply testing whether the return type is consistent with the input type, we also test the converse. If the two types are not the same and they are consistent in both directions, then we cannot use strict type guard semantics. If they are the same or they are consistent in only one direction, then we can use strict type guard semantics.
2. If the input type contains any types that do not follow subtyping rules (i.e. it contains an `Any` or a `...`, even in a nested type (e.g. `list[Any]` or `list[list[Any]] or `list[Callable[..., None]]`, then we cannot use strict type guard semantics.

These two options are slightly different, and I need to give more thought to which one is "more correct" from a type theory perspective. 

I'll note that either of these options would fix this issue and the case described in #5815, which uses `type[Any]` as an input type.

**Comment by erictraut** [comment]
This is included in pyright 1.1.325, which I just published. It will also be included in an upcoming pylance release.
